lib/plugins/browser-sync.js

```diff
@@ -0,0 +1,62 @@
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+'use strict';
+
+const loaderFeatures = require('../features');
+
+/**
+ * @param {Array} plugins
+ * @param {WebpackConfig} webpackConfig
+ * @returns {void}
+ */
+module.exports = function(plugins, webpackConfig) {
+
+    if (!webpackConfig.useBrowserSync) {
+        return;
+    }
+
+    loaderFeatures.ensurePackagesExist('browsersync');
+    const BrowserSyncPlugin = require('browser-sync-webpack-plugin');
+
+    let browserSyncOptions = {};
+    let browserSyncPluginOptions = {};
+    // allow to overwrite values by user
+    webpackConfig.browserSyncOptionsCallback.apply(null,
+        [browserSyncOptions, browserSyncPluginOptions]
+    );
+
+    let url = require('url');
+    let serverParts = url.parse(webpackConfig.browserSyncOptions.backendUrl, true);
+    let browserSyncConfig = Object.assign({
+        proxy: webpackConfig.browserSyncOptions.backendUrl,
+        port: serverParts.port || '80',
```

Comment:
I'm not sure to understand why you use the port from `backendUrl` here.

Shouldn't `proxy` contain the URL of the dev server (so something like `http://localhost:8080/` by default) and `port` the port Browsersync will listen to? Is there a case in which the port can be the same in both of them?

Also, as I said previously I don't think the `backendUrl` option is needed anyway. Removing it would also remove the need of the `url` dependency (which may look a bit strange and out-of-place for users).

Reply:
Regarding the `port` settings

I think we must specify it at the moment we use the `proxy` option, the [default is 3000](https://browsersync.io/docs/options#option-port) so not setting `port` could not work am afraid. Although to be honest I need to test it, did not think on that before.

I did decide to use `proxy` option because it cover more general configuration use case according to [plugin](https://github.com/Va1/browser-sync-webpack-plugin) use alongside with webpack-dev server and so on.

Regarding `backendUrl` to complement what said in previous comment, to remove `url` deps we could also pass port as parameter and them build proxy url concatenating both values. Signature would be:
```
Encore.enableBrowserSync('http://localhost', '8080', ['./*.twig', './*.php'],  
    function(browserSyncOptions, pluginOptions) {}
);
```
this of course in case we continue using parameters. If only callback used, then we don't need anything else.

Reply:
I definitely agree about using `proxy` but according to the plugin page you linked the port in this case should be the one used by the dev-server (I don't know if we can retrieve it at this point, that would be great though...), whereas the one in the `port` property represents the port browser-sync will listen to:

```js
{
  // browse to http://localhost:3000/ during development
  host: 'localhost',
  port: 3000,
  // proxy the Webpack Dev Server endpoint
  // (which should be serving on http://localhost:3100/)
  // through BrowserSync
  proxy: 'http://localhost:3100/'
},
```

If you use `backendUrl` both of them will end-up having the same value which doesn't seem right.

Reply:
I think we are mixing approaches from plugin page. Here in current functionality, `host` option is not used. Only `proxy`, also, you made me spot a mistake about my approach, thanks! The `port` option is not required at all and nothing would brake if not specified, sorry for stating that before. I had the `port` option as a mean to run several project at once without conflicts. 

So for example (I am not using port here, so default 3000 is used) by using the feature like this:
```
Encore.enableBrowserSyncPlugin('http://localhost:8091', [paths]);
```
this mean I have my php server running on: `http://localhost:8091`, which is the url I want to proxy with brower sync, in a separate terminal I ran: `php -S 127.0.0.1:8091`. Then in another terminal when I run `encore dev-server` I get an output like:
```
> encore dev-server

Running webpack-dev-server ...

Project is running at http://localhost:8080/
webpack output is served from /static/theme/
Content not from webpack is served from /home/shak/workspace/www
404s will fallback to /index.html
 DONE  Compiled successfully in 66501ms                                                                                                       1:28:25 PM

[BS] Proxying: http://127.0.0.1:8091
[BS] Access URLs:
 -------------------------------------
       Local: http://localhost:3000
    External: http://192.168.0.26:3000
 -------------------------------------
          UI: http://localhost:3001
 UI External: http://192.168.0.26:3001
 -------------------------------------
[BS] Watching files...
```

>Project is running at http://localhost:8080/
>webpack output is served from /static/theme/

this means that if you browse: `http://127.0.0.1:8091/static/theme/some-css-file.css` webpack dev server will give you the in memory compilation it has currently.

> [BS] Proxying: http://127.0.0.1:8091

this is, as before out back end (php) server url, even if you don't use browser sync you can access this url.

> [BS] Access URLs:
>       Local: http://localhost:3001
>    External: http://192.168.0.26:3001

these are the urls served by browser sync so you can have the live reloading as well as browser synchronization features, when you browse one of those, you are accessing your backend through browser sync, which in turn is watching the directories you gave before and when a change occur all browsers browsing that url will reload and show updated page.

does this explain better what functionality does currently?

Reply:
> Here in current functionality, host option is not used

It is, just with the "[guessed](https://browsersync.io/docs/options#option-host)" value (probably `192.168.0.26` based on your logs).

> this means that if you browse: http://127.0.0.1:8091/static/theme/some-css-file.css webpack dev server will give you the in memory compilation it has currently.

That doesn't feel right. If you query `127.0.0.1:8091` browser-sync won't be reached since your PHP server is using that port.

Currently it seems that you have:

```
   - browser-sync (3000) ---[proxy]---> php server (8091)
--|
   - dev-server (8080)
```

I thought you were trying to implement [this part](https://github.com/Va1/browser-sync-webpack-plugin#advanced) of the documentation, which would basically look like:

```
-- browser-sync (3000) ---[proxy]---> dev-server (8080)
```

That also explains why you also added the "watch" part and why you would need to set the proxy option everytime you call `Encore.enableBrowserSync()`. However I'm wondering if that would be the most common use case of BrowserSync.

Let's say I'm an user working on a React project consuming an external API (so no PHP side-project). I see that there is an `Encore.enableBrowserSync()` method, what would I expect it to do?

Reply:
>Let's say I'm an user working on a React project consuming an external API (so no PHP side-project). I see that there is an Encore.enableBrowserSync() method, what would I expect it to do?

I see you point. I wanted to discover scenarios like this with this issue. I am not a React developer. Please could you explain how would be the use case and what would you expect from the API. Also please could you propose an API for it.

>I thought you were trying to implement this part of the documentation

You are right, it was that one. Wondering now if I got it right? :)

To resume. I am having the feeling now that I implemented only an specific use case for the tool. If other uses cases deviate from current solution I think your initial proposition of having the option callback only would be the general common denominator. Not sure if worth the effort to implement more than that. 

Thinking now whether this browser-sync integration issue could be best described in documentation, through snippets maybe, on how to use for the different uses cases.

Reply:
I'm not really a React developer either, but the same idea applies to any other framework (Angular, Vue, ...).

I quickly added Browsersync to my preact demo to show what I meant by that: https://github.com/Lyrkan/encore-preact/tree/browser-sync

If you look at the `webpack.config.js` file you'll see that the value of `proxy` is the URL of the dev-server (unless you use `--port`). So when you run `yarn encore dev-server` you'll be able to browse `http://localhost:3000/` and have access to both the dev-server (since it is proxied) and the features from Browsersync (events mirroring between browsers, network throttling, debug tools, etc.).